community/kernel/src/main/java/org/neo4j/kernel/impl/traversal/BidirectionalTraverserIterator.java

```diff
@@ -91,7 +73,26 @@ public Side( MonoDirectionalTraversalDescription description )
                 new AsOneStartBranch( this, endNodes, end.initialState, start.uniqueness ), end.expander );
 
         this.selector = sideSelector.create( startSelector, endSelector, maxDepth );
-        this.collisionDetector = collisionPolicy.create( collisionEvaluator );
+        this.collisionDetector = buildCollisionDetector( collisionPolicy, collisionEvaluator );
+    }
+
+    private BranchCollisionDetector buildCollisionDetector(
+            org.neo4j.graphdb.traversal.BranchCollisionPolicy collisionPolicy,
+            PathEvaluator collisionEvaluator )
+    {
+        BranchCollisionDetector branchCollisionDetector = collisionPolicy.create( collisionEvaluator );
+        if ( branchCollisionDetector instanceof Filterable )
```

Comment:
You'd only need to change `BranchCollisionPolicy` interface, right? I remember an old discussion about things like these, especially regarding traversal API. What I think you can do is to add a method `BranchCollisionDetector create( Evaluator evaluator )` in `BranchCollisionPolicy` as well as an `Adapter` a la `PathEvaluator` to bridge the change. Not sure we even expect users to have implementations of `BranchCollisionPolicy`.